src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculator.java

```diff
@@ -0,0 +1,167 @@
+package uk.gov.homeoffice.digital.sas.balancecalculator;
+
+import com.google.common.collect.Range;
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.gov.homeoffice.digital.sas.balancecalculator.client.RestClient;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.Accrual;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.Agreement;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.Contributions;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.enums.AccrualType;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.timecard.TimeEntry;
+
+@Component
+public class BalanceCalculator {
+
+  private final RestClient restClient;
+
+  @Autowired
+  public BalanceCalculator(RestClient restClient) {
+    this.restClient = restClient;
+  }
+
+  public List<Accrual> calculate(TimeEntry timeEntry) {
+
+    List<Accrual> accrualsToBatchUpdate = Arrays.stream(AccrualType.values())
+        .filter(AccrualType::isEnabled)
+        .map(accrualType -> calculateAccruals(timeEntry, accrualType))
+        .flatMap(Collection::stream)
+        .toList();
+
+    // TODO : send Batch Update request to Accruals API
+    // restClient.batchUpdate(accrualsToBatchUpdate)
+
+    return accrualsToBatchUpdate;
+  }
+
+  List<Accrual> calculateAccruals(TimeEntry timeEntry, AccrualType accrualType) {
+
+    Map<LocalDate, Range<ZonedDateTime>> dateRangeMap =
+        splitOverDays(timeEntry.getActualStartTime(), timeEntry.getActualEndTime());
+
+    return dateRangeMap.entrySet().stream()
+        .map(entry -> recalculateContributions(timeEntry.getId(), timeEntry.getTenantId(),
+            timeEntry.getOwnerId(), accrualType, entry.getKey(), entry.getValue()))
+        .map(accrual -> cascadeCumulativeTotal(timeEntry, accrual)).flatMap(Collection::stream)
+        .toList();
+  }
+
+  private List<Accrual> cascadeCumulativeTotal(TimeEntry timeEntry, Accrual accrual) {
+    LocalDate referenceDate = accrual.getAccrualDate();
+
+    List<Accrual> accrualsToUpdate =
+        getAccrualsFromReferenceDateUntilEndOfAgreement(timeEntry.getTenantId(),
+            timeEntry.getOwnerId(), accrual.getAgreementId().toString(), referenceDate.plusDays(1));
+    accrualsToUpdate.add(0, accrual);
+
+    BigDecimal priorCumulativeTotal =
+        restClient.getPriorAccrual(timeEntry.getTenantId(), timeEntry.getOwnerId(),
```

Comment:
I wonder if a separate method on the restAPI is needed here. If we were to get the accruals up front we could always call `getAccrualsBetweenDates(..... startDate - 1, endDate, accrualType)`

Reply:
Actioned. Accruals of all types are fetched between startDate -1 and end of agreement (lines 53-55 in BalanceCalculator`):

`    Map<AccrualType, TreeMap<LocalDate, Accrual>> allAccruals =
        getAccrualsBetweenDates(tenantId, personId,
            timeEntryStartDate.minusDays(1), applicableAgreement.getEndDate());
`